include/curl/easy.h

```diff
@@ -25,6 +25,98 @@
+ *  with OpenSSL, Schannel or SecTransport
+ *
+ *
+ *  Example if with certificate binary data and size in:
+ *    void* certdata;
+ *    size_t certsize;
+ *    ...
+ *  set the certificate with:
+ *    struct curl_blob structblob;
+ *    curl_init_blob_dup(&structblob, certdata, certsize);
+ *    curl_easy_setopt(curl, CURLOPT_KEYPASSWD, &structblob);
+ *             my_setopt_str(curl, CURLOPT_SSLCERT, &structblob);
+ *  struct curl_blob is just a 0x20 bytes structures which will
+ *  start bt a magic string and contain data
+ *
+ *  with curl_init_blob_dup, certdata can be discarded after
+ *    calling curl_easy_setopt
+ *  with curl_init_blob_persist, certdata need be valid when the
+ *  connexion is made
+ *
+ * Note : never call curl_init_blob_xxx without a pointer to a
+ *   valid 0x20 bytes block
+ * And never use the string CURL_BLOB_MAGIC without a correct
+ *   blob block
+ *
+ *
+ * about the idea, see https://curl.haxx.se/mail/lib-2016-09/0074.html */
+
+#define CURL_BLOB_MAGIC          ("\x01" "CurlMemBlob" "\xff\x01")
+#define CURL_BLOB_MAGIC_STRLEN   (14)
+#define CURL_BLOB_MAGIC_SIZE     (CURL_BLOB_MAGIC_STRLEN + 1)
+#define CURL_BLOB_OFFSET_DUPFLAG (CURL_BLOB_MAGIC_SIZE)
+#define CURL_BLOB_SIZE_DUPFLAG   (1)
+#define CURL_BLOB_OFFSET_DATALEN (CURL_BLOB_OFFSET_DUPFLAG + \
+                                  CURL_BLOB_SIZE_DUPFLAG)
+#define CURL_BLOB_SIZE_DATALEN   (sizeof(size_t))
+#define CURL_BLOB_OFFSET_DATAPTR (CURL_BLOB_OFFSET_DATALEN + \
+                                  CURL_BLOB_SIZE_DATALEN)
+#define CURL_BLOB_SIZE_DATAPTR   (sizeof(void *))
+#define CURL_BLOB_SIZE           (CURL_BLOB_OFFSET_DATAPTR + \
+                                  CURL_BLOB_SIZE_DATAPTR)
+
+#define CURL_BLOB_DUPFLAG_COPY   (1)
+#define CURL_BLOB_DUPFLAG_NOCOPY (0)
+
+/* the struct curl_blob store binary data parameters
+ * the structure size is 0x20 and can be followed by binary data */
+struct curl_blob {
+  char blob_internal[CURL_BLOB_SIZE];
+};
+
+
+/*
+ * Macro WHILE_FALSE may be used to build single-iteration do-while loops,
+ * avoiding compiler warnings. Mostly intended for other macro definitions.
+ */
+
+#define WHILE_FALSE_EASYCURL  while(0)
+
+#if defined(_MSC_VER) && !defined(__POCC__)
+#  undef WHILE_FALSE_EASYCURL
+#  if (_MSC_VER < 1500)
+#    define WHILE_FALSE_EASYCURL  while(1, 0)
+#  else
+#    define WHILE_FALSE_EASYCURL \
+__pragma(warning(push)) \
+__pragma(warning(disable:4127)) \
+while(0) \
+__pragma(warning(pop))
+#  endif
+#endif
+
+#define curl_init_blob_flag(structblob, ptr, len, flag) \
+  do { \
+    char *blob = (structblob)->blob_internal; \
+    char *data = (char *)(ptr); \
+    size_t size = (size_t)(len); \
+    memcpy(blob, CURL_BLOB_MAGIC, CURL_BLOB_MAGIC_SIZE); \
+    memcpy(blob + CURL_BLOB_OFFSET_DATALEN, &size, sizeof(size_t)); \
+    memcpy(blob + CURL_BLOB_OFFSET_DATAPTR, &data, sizeof(void *)); \
+    blob[CURL_BLOB_OFFSET_DUPFLAG] = (char)(flag); \
+  } WHILE_FALSE_EASYCURL
```

Comment:
Using the plain struct instead, this can be simplified a lot. Possibly it isn't even needed?